hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bloom/SparkHoodieBloomIndex.java

```diff
@@ -122,13 +122,15 @@ public SparkHoodieBloomIndex(HoodieWriteConfig config) {
 
     // Step 3: Obtain a RDD, for each incoming record, that already exists, with the file id,
     // that contains it.
+    JavaRDD<Tuple2<String, HoodieKey>> fileComparisonsRDD =
```

Comment:
do you actually see from the Spark UI that its not computed twice? I ask because, `fileComparisonsRDD` is not cached and thus even though this is declared only once, during runtime, Spark will lazily recompute `fileComparisonsRDD` once for each method that uses it.

Reply:
I didn't check the Spark UI yet, just a simple analyze the process of data writing. For each batch of records to write, the `SparkHoodieBloomIndex.lookupIndex` was expected to be invoked once so the `fileComparisonsRDD` should only be evaluated only once, is there other invocation for `SparkHoodieBloomIndex.lookupIndex` ? Maybe i missed something.

Reply:
So Spark does lazy evaluation of an RDD. If the RDD is not persisted to disk/cache, it simply recomputes it. In this case, `fileComparisonsRDD` would be recomputed twice during runtime. the method `explodeRecordRDDWithFileComparisons()` will only be called once, but it does nothing in practice except "define" the RDD it returns. 

In contrast, if you notice this line of code at the start of `tagLocation`

```
// Step 0: cache the input record RDD
    if (config.getBloomIndexUseCaching()) {
      recordRDD.persist(SparkMemoryUtils.getBloomIndexInputStorageLevel(config.getProps()));
    }
```

This caches the incoming recordRDD and thus however many times this RDD is used in the indexing DAG, it will not go to the previous stage. if we did not have the .persist() in here, then everytime an RDD derived off this `recordRDD` is needed for a Spark action, it will keep reading from source and compute the recordRDD again. 

Apologies, if you knew all this already. :) and I am failing to see how this wont happen. but at-least you get my concern with this explanation.

Reply:
Thanks for the explanation, the `recordRDD` may be persisted, but the computation in `explodeRecordRDDWithFileComparisons` still need to do 2 times, right ?

Reply:
Correct. That's what i think will happen